[16.0][MIG] product_cost_security: Migration to 16.0

bvanjeelharia opened the pull request.

sergio-teruel committed:
[11.0][ADD] product_cost_security: New module to restrict product cost view

sergio-teruel committed:
[11.0][ADD] product_cost_security: Fix translate folder name

sergio-teruel committed:
[11.0][IMP] product_cost_security: Avoid use odoo PR 28755
https://github.com/odoo/odoo/pull/28755

oca-travis committed:
[UPD] Update product_cost_security.pot

OCA-git-bot committed:
[UPD] README.rst

sergio-teruel committed:
[MIG] product_cost_security: Migration to v12.0

oca-travis committed:
[UPD] Update product_cost_security.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: product-attribute-12.0/product-attribute-12.0-product_cost_security
Translate-URL: https://translation.odoo-community.org/projects/product-attribute-12-0/product-attribute-12-0-product_cost_security/

sergio-teruel committed:
[IMP] product_cost_security: black, isort

sergio-teruel committed:
[MIG] product_cost_security: Migration to v13.0

oca-travis committed:
[UPD] Update product_cost_security.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: product-attribute-13.0/product-attribute-13.0-product_cost_security
Translate-URL: https://translation.odoo-community.org/projects/product-attribute-13-0/product-attribute-13-0-product_cost_security/

OCA-git-bot committed:
pre-commit update

sergio-teruel committed:
[IMP] product_cost_security: Add user_root to security group

OCA-git-bot committed:
product_cost_security 13.0.1.1.0

watthanun committed:
[MIG] product_cost_security: Migration to 14.0

oca-travis committed:
[UPD] Update product_cost_security.pot

OCA-git-bot committed:
[UPD] README.rst

Mat-moran committed:
Translated using Weblate (Spanish)

Currently translated at 80.0% (4 of 5 strings)

Translation: product-attribute-14.0/product-attribute-14.0-product_cost_security
Translate-URL: https://translation.odoo-community.org/projects/product-attribute-14-0/product-attribute-14-0-product_cost_security/es/

osoul-dev committed:
Added translation using Weblate (Arabic)

osoul-dev committed:
Translated using Weblate (Arabic)

Currently translated at 87.5% (7 of 8 strings)

Translation: product-attribute-14.0/product-attribute-14.0-product_cost_security
Translate-URL: https://translation.odoo-community.org/projects/product-attribute-14-0/product-attribute-14-0-product_cost_security/ar/

celm1990 committed:
[MIG] product_cost_security: Migration to 15.0
* split models per file

Commit:
[UPD] Update product_cost_security.pot

OCA-git-bot committed:
[UPD] README.rst

Commit:
[UPD] Update product_cost_security.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: product-attribute-15.0/product-attribute-15.0-product_cost_security
Translate-URL: https://translation.odoo-community.org/projects/product-attribute-15-0/product-attribute-15-0-product_cost_security/

pablontura committed:
Added translation using Weblate (Catalan)

pablontura committed:
Translated using Weblate (Catalan)

Currently translated at 80.0% (4 of 5 strings)

Translation: product-attribute-15.0/product-attribute-15.0-product_cost_security
Translate-URL: https://translation.odoo-community.org/projects/product-attribute-15-0/product-attribute-15-0-product_cost_security/ca/

sergio-teruel committed:
[IMP] product_cost_security: Add a security group to modify the cost price.

Commit:
[UPD] Update product_cost_security.pot

OCA-git-bot committed:
product_cost_security 15.0.2.0.0

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: product-attribute-15.0/product-attribute-15.0-product_cost_security
Translate-URL: https://translation.odoo-community.org/projects/product-attribute-15-0/product-attribute-15-0-product_cost_security/

astirpe commented on product_cost_security/views/product_views.xml:
```diff
@@ -6,21 +6,23 @@
-                />
+                    name="attrs"
+                >{'invisible': [('user_can_see_standard_price_uom', '=', False),'|',('product_variant_count', '&gt;',
+                    1), ('is_product_variant', '=', False)]}
```
This condition could be moved into `_compute_user_can_see_standard_price_uom()`

astirpe commented on product_cost_security/tests/test_product_cost_security.py:
```diff
@@ -0,0 +1,69 @@
+                [self.product_edit_cost_group_id],
+            )
+        ]
+        sheet_form = Form(self.product_template_model.with_user(self.env.user))
```
You are testing three different cases here, so you could split this test and have 3 tests in total

bvanjeelharia replied:
Hello @astirpe , I have made the changes as per your suggestions.There are now 7 tests in all, as I had to add a field 'user_can_see_standard_price_uom' for product.product (product variant) as well.

Thanks

astirpe commented:
You squashed the V16 changes into the pre-commit

bvanjeelharia committed:
[IMP] product_cost_security: pre-commit stuff

bvanjeelharia committed:
[MIG] product_cost_security: Migration to 16.0

[MIG] product_cost_security: Migration to 16.0

[MIG] product_cost_security: Migration to 16.0

[MIG] product_cost_security: Migration to 16.0

[MIG] product_cost_security: Migration to 16.0

Updated test scripts

astirpe commented on product_cost_security/models/product_template.py:
```diff
@@ -17,3 +20,11 @@ def _compute_user_can_update_cost(self):
+        for product in self:
+            product.user_can_see_standard_price_uom = self.env.user.has_group(
+                "product_cost_security.group_product_cost"
+            ) and (product.product_variant_count <= 1 or product.is_product_variant)
```
This method should depend on `"product_variant_count"` and `"is_product_variant"`

bvanjeelharia replied:
Yes, I will make that change

astirpe commented on product_cost_security/models/product_product.py:
```diff
@@ -1,9 +1,20 @@
     _inherit = "product.product"
 
     standard_price = fields.Float(groups="product_cost_security.group_product_cost")
+    user_can_see_standard_price_uom = fields.Boolean(
```
Why did you add this field? Is not used in the product.product view

bvanjeelharia replied:
Its because the view is common for both product.product and product.template

bvanjeelharia replied:
and the field was getting calculated based on product.template instead of product.product

astirpe commented on product_cost_security/views/product_views.xml:
```diff
@@ -6,21 +6,22 @@
                 <attribute
-                    name="groups"
-                    separator=","
-                    add="product_cost_security.group_product_cost"
```
In the V15 version, group `product_cost_security.group_product_cost` was added to already existing "`base.group_user`"  for this field. So in theory you may have multiple groups for `standard_price_uom`.
But in `_compute_user_can_see_standard_price_uom()` you only handle one group. Why did you add `product_variant_count` and `is_product_variant` in the condition? Can't you just leave things as they were in V15?

bvanjeelharia replied:
Sorry for the inconveniences caused, this happened as the Odoo base addons were recently changed(https://github.com/odoo/odoo/commit/206e88b0ad9fdb6e1767c93b3eca8feeecddd3da) and I had an old db where the group was still applied on the field and it was causing an additional field (uom) to show up, please check attached screenshot where in the view is disturbed with per Units showing up which should have hidden. I have restored it to how it was in v15
![Screenshot from 2023-01-03 16-50-57](https://user-images.githubusercontent.com/88703470/210347866-f9622b41-9bec-4215-82aa-a8a506600e4b.png)

bvanjeelharia commented:
Can someone please review and approve this PR?

rafaelbn commented:
Please @bvanjeelharia , could you make a "commit" empty to try un block runboat in order to test it?

@sergio-teruel you could be interested in this PR 😄

sergio-teruel commented:
@rafaelbn please give me the task :smile:

sergio-teruel commented on product_cost_security/tests/test_product_cost_security.py:
```diff
@@ -0,0 +1,77 @@
+class TestProductCostSecurity(TransactionCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
```
You can create an environment with tracking disable.
cls.env = cls.env(context=dict(cls.env.context, tracking_disable=True))

bvanjeelharia replied:
Made the change

sergio-teruel commented on product_cost_security/tests/test_product_cost_security.py:
```diff
@@ -0,0 +1,77 @@
+            standard_price = sheet_form.standard_price
+            self.assertEqual(standard_price, 0.0)
+
+    def test_1(self):
```
Can you use more descriptive names?

bvanjeelharia replied:
Yes, I have made the changes

sergio-teruel reviewed (approved):
Minor changes
Please, can you modify the commit deleting the extra lines?

CasVissers-360ERP commented:
@rafaelbn can be merged?

rafaelbn reviewed (approved):
Thank you! 👍

bvanjeelharia commented:
Can this be merged now?

pedrobaeza commented:
/ocabot migration product_cost_security
/ocabot merge nobump

Yadier-Tecnativa reviewed (commented):
check
